File: src/main/java/jukebot/audio/AudioHandler.java
package jukebot.audio;

import com.sedmelluq.discord.lavaplayer.filter.equalizer.EqualizerFactory;
import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter;
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason;
import com.sedmelluq.discord.lavaplayer.track.playback.AudioFrame;
import jukebot.Database;
import jukebot.JukeBot;
import jukebot.utils.Helpers;
import jukebot.utils.Permissions;
import net.dv8tion.jda.core.EmbedBuilder;
import net.dv8tion.jda.core.audio.AudioSendHandler;
import net.dv8tion.jda.core.entities.Guild;
import net.dv8tion.jda.core.entities.TextChannel;
import net.dv8tion.jda.core.managers.AudioManager;

import java.util.HashSet;
import java.util.LinkedList;
import java.util.Random;
import java.util.concurrent.TimeUnit;

public class AudioHandler extends AudioEventAdapter implements AudioSendHandler {

    private final Permissions permissions = new Permissions();

    public AudioPlayer player;
    private EqualizerFactory equalizer = new EqualizerFactory();

    private AudioFrame lastFrame;
    private final Random selector = new Random();
    private bassBoost bassBoostMode = AudioHandler.bassBoost.OFF;

    private final LinkedList<AudioTrack> queue = new LinkedList<>();
    private final HashSet<Long> skips = new HashSet<>();
    private Long channelId;
    private Long guildId;
    private boolean shouldAnnounce = true;

    private repeatMode repeat = repeatMode.NONE;
    private boolean shuffle = false;

    private AudioTrack current = null;

    public int trackPacketLoss = 0;
    public int trackPackets = 0;

    public AudioHandler(Long guildId, AudioPlayer player) {
        this.guildId = guildId;
        this.player = player;
        player.addListener(this);
    }

    public boolean enqueue(AudioTrack track, Long userID) { // boolean: shouldAnnounce
        track.setUserData(userID);

        if (!player.startTrack(track, true)) {
            queue.offer(track);
            return true;
        }

        return false;
    }

    public boolean isPlaying() {
        return player.getPlayingTrack() != null;
    }

    public boolean isBassBoosted() {
        return bassBoostMode != bassBoost.OFF;
    }

    public LinkedList<AudioTrack> getQueue() {
        return queue;
    }

    public String getRepeatMode() {
        return repeat.toString().toLowerCase();
    }

    public String getBassBoostSetting() {
        return bassBoostMode.toString().toLowerCase();
    }

    public boolean isShuffleEnabled() {
        return shuffle;
    }

    public boolean toggleShuffle() {
        return shuffle = !shuffle;
    }

    public void setRepeat(repeatMode mode) {
        repeat = mode;
    }

    public int voteSkip(Long userID) {
        skips.add(userID);
        return skips.size();
    }

    public void setChannel(Long channelId) {
        this.channelId = channelId;
    }

    public void setShouldAnnounce(Boolean shouldAnnounce) {
        this.shouldAnnounce = shouldAnnounce;
    }

    public void playNext() {
        AudioTrack nextTrack = null;

        if (repeat == repeatMode.ALL && current != null) {
            AudioTrack r = current.makeClone();
            r.setUserData(current.getUserData());
            queue.offer(r);
        }

        if (repeat == repeatMode.SINGLE && current != null) {
            nextTrack = current.makeClone();
            nextTrack.setUserData(current.getUserData());
        } else if (!queue.isEmpty()) {
            nextTrack = shuffle ? queue.remove(selector.nextInt(queue.size())) : queue.poll();
        }

        if (nextTrack != null) {
            player.startTrack(nextTrack, false);
        } else {
            current = null;
            player.stopTrack();
            bassBoost(bassBoost.OFF);

            final Guild guild = JukeBot.shardManager.getGuildById(guildId);

            if (guild == null) { // Bot was kicked or something
                JukeBot.removePlayer(guildId);
                return;
            }

            final AudioManager audioManager = guild.getAudioManager();

            if (audioManager.isConnected() || audioManager.isAttemptingToConnect()) {
                Helpers.Companion.schedule(audioManager::closeAudioConnection, 1, TimeUnit.SECONDS);

                announce("Queue Concluded!", "[Support the development of JukeBot!](https://www.patreon.com/Devoxin)" +
                        "\nSuggest features with the `feedback` command!");
            }
        }
    }

    private void announce(String title, String description) {
        if (!shouldAnnounce || channelId == null) {
            return;
        }

        final TextChannel channel = JukeBot.shardManager.getTextChannelById(channelId);

        if (channel == null || !permissions.canSendTo(channel)) {
            return;
        }

        channel.sendMessage(new EmbedBuilder()
                .setColor(Database.getColour(channel.getGuild().getIdLong()))
                .setTitle(title)
                .setDescription(truncate(description, 1000))
                .build()
        ).queue(null, err -> JukeBot.LOG.error("Encountered an error while posting track announcement", err));
    }

    private String truncate(String content, int maxLength) {
        if (content.length() > maxLength) {
            return content.substring(0, maxLength - 3) + "...";
        }

        return content;
    }

    public void cleanup() {
        queue.clear();
        skips.clear();
        player.destroy();

        final Guild g = JukeBot.shardManager.getGuildById(guildId);
        if (g != null) {
            g.getAudioManager().setSendingHandler(null);
        }
    }

    public void bassBoost(bassBoost preset) {
        if (preset != bassBoost.OFF && bassBoostMode == bassBoost.OFF) {
            player.setFilterFactory(equalizer);
        } else if (preset == bassBoost.OFF && bassBoostMode != bassBoost.OFF) {
            player.setFilterFactory(null);
        }

        bassBoostMode = preset;
        equalizer.setGain(0, preset.band0);
        equalizer.setGain(1, preset.band1);
        equalizer.setGain(2, preset.band2);
    }

    /*
     * Player Events
     */

    @Override
    public void onTrackEnd(AudioPlayer player, AudioTrack track, AudioTrackEndReason endReason) {
        skips.clear();
        trackPacketLoss = 0;
        trackPackets = 0;

        if (endReason.mayStartNext)
            playNext();
    }

    @Override
    public void onTrackStart(AudioPlayer player, AudioTrack track) {
        player.setPaused(false);

        if (current == null || !current.getIdentifier().equals(track.getIdentifier())) {
            current = track;
            announce("Now Playing", track.getInfo().title);
        }
    }

    @Override
    public void onTrackException(AudioPlayer player, AudioTrack track, FriendlyException exception) {
        if (repeat != repeatMode.NONE)
            repeat = repeatMode.NONE;

        announce("Playback Error", "Playback of **" + track.getInfo().title + "** encountered an error!\n" +
                exception.getLocalizedMessage());
    }

    @Override
    public void onTrackStuck(AudioPlayer player, AudioTrack track, long thresholdMs) {
        if (repeat != repeatMode.NONE)
            repeat = repeatMode.NONE;

        announce("Track Stuck", "Playback of **" + track.getInfo().title + "** has frozen and is unable to resume!");
        playNext();
    }

    /*
     * Packet Sending Events
     */

    @Override
    public boolean canProvide() {
        lastFrame = player.provide();

        if (!player.isPaused()) {
            if (lastFrame == null) {
                trackPacketLoss++;
            } else {
                trackPackets++;
            }
        }

        return lastFrame != null;
    }

    @Override
    public byte[] provide20MsAudio() {
        return lastFrame.getData();
    }

    @Override
    public boolean isOpus() {
        return true;
    }

    public enum repeatMode {
        SINGLE, ALL, NONE
    }

    public enum bassBoost {
        OFF(0F, 0F, 0F),
        WEAK(0.03F, 0.01F, 0.0F),
        MEDIUM(0.1F, 0.08F, 0.04F),
        STRONG(0.2F, 0.15F, 0.11F),
        INSANE(0.4F, 0.26F, 0.18F),
        WTF(1F, 0.8F, 0.6F);

        private final float band0;
        private final float band1;
        private final float band2;

        bassBoost(float b0, float b1, float b2) {
            this.band0 = b0;
            this.band1 = b1;
            this.band2 = b2;
        }
    }

}

File: src/main/java/jukebot/audio/AudioHandler.kt
package jukebot.audio

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason
import com.sedmelluq.discord.lavaplayer.track.playback.MutableAudioFrame
import io.sentry.Sentry
import io.sentry.event.BreadcrumbBuilder
import io.sentry.event.Event
import io.sentry.event.EventBuilder
import io.sentry.event.interfaces.ExceptionInterface
import jukebot.Database
import jukebot.JukeBot
import jukebot.utils.Helpers
import jukebot.utils.toTimeString
import jukebot.utils.toTitleCase
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.audio.AudioSendHandler
import java.nio.ByteBuffer
import java.util.*
import java.util.concurrent.TimeUnit

class AudioHandler(private val guildId: Long, val player: AudioPlayer) : AudioEventAdapter(), AudioSendHandler {
    private val mutableFrame = MutableAudioFrame()
    private val buffer = ByteBuffer.allocate(1024)
    // ByteBuffer.allocate(StandardAudioDataFormats.DISCORD_OPUS.maximumChunkSize())

    // Playback Settings
    val bassBooster = BassBooster(player)
    var repeat = RepeatMode.NONE
    var shuffle = false

    val queue = LinkedList<AudioTrack>()
    private val skips = hashSetOf<Long>()
    private val selector = Random()

    var shouldAnnounce = true
    var channelId: Long? = null

    // Performance Tracking
    var trackPacketLost = 0
    var trackPacketsSent = 0

    // Player Stuff
    val autoPlay = AutoPlay(guildId)
    var previous: AudioTrack? = null
    var current: AudioTrack? = null
    val isPlaying: Boolean
        get() = player.playingTrack != null

    init {
        player.addListener(this)
        this.mutableFrame.setBuffer(buffer)
    }

    fun enqueue(track: AudioTrack, userID: Long, playNext: Boolean): Boolean { // boolean: shouldAnnounce
        track.userData = userID

        if (!player.startTrack(track, true)) {
            if (playNext) {
                queue.add(0, track)
            } else {
                queue.add(track)
            }
            return true
        }

        return false
    }

    fun voteSkip(userID: Long): Int {
        skips.add(userID)
        return skips.size
    }

    fun playNext(shouldAutoPlay: Boolean = true) {
        var nextTrack: AudioTrack? = null

        current?.let {
            if (it.sourceManager.sourceName == "youtube") {
                autoPlay.store(it.info.title)
            }
        }

        if (current != null && repeat != RepeatMode.NONE) {
            val cloned = current!!.makeClone().also { c -> c.userData = current!!.userData }

            if (repeat == RepeatMode.ALL) {
                queue.offer(cloned)
            } else if (repeat == RepeatMode.SINGLE) {
                nextTrack = cloned
            }
        }

        if (nextTrack == null && !queue.isEmpty()) {
            nextTrack = if (shuffle) queue.removeAt(selector.nextInt(queue.size)) else queue.poll()
        }

        if (nextTrack != null) {
            return player.playTrack(nextTrack)
        }

        if (shouldAutoPlay && autoPlay.enabled && autoPlay.hasSufficientData) {
            autoPlay.getRelatedTrack()
                .thenAccept(player::playTrack)
                .exceptionally {
                    playNext(false)
                    announce("AutoPlay", "AutoPlay encountered an error.\nWe're sorry for any inconvenience caused!")
                    Sentry.capture(it)
                    return@exceptionally null
                }
            return
        }

        current = null
        player.stopTrack()
        bassBooster.boost(0.0f)

        val audioManager = JukeBot.shardManager.getGuildById(guildId)?.audioManager
            ?: return JukeBot.removePlayer(guildId)

        if (audioManager.isConnected) {
            Helpers.schedule(audioManager::closeAudioConnection, 1, TimeUnit.SECONDS)

            if (Database.getIsPremiumServer(guildId)) {
                announce("Queue Concluded", "Enable AutoPlay to keep the party going!")
            } else {
                announce("Queue Concluded!",
                    "[Support the development of JukeBot!](https://www.patreon.com/Devoxin)")
            }

            setNick(null)
        }
    }

    private fun announce(title: String, description: String) {
        if (!shouldAnnounce || channelId == null) {
            return
        }

        val channel = JukeBot.shardManager.getTextChannelById(channelId!!)
            ?.takeIf { Helpers.canSendTo(it) }
            ?: return

        channel.sendMessage(EmbedBuilder()
            .setColor(Database.getColour(channel.guild.idLong))
            .setTitle(title)
            .setDescription(Helpers.truncate(description, 1000))
            .build()
        ).queue()
    }

    private fun setNick(nick: String?) {
        val guild = JukeBot.shardManager.getGuildById(guildId) ?: return

        if (guild.selfMember.hasPermission(Permission.NICKNAME_CHANGE) && Database.getIsMusicNickEnabled(guild.idLong)) {
            guild.selfMember.modifyNickname(nick).queue()
        }
    }

    fun cleanup() {
        queue.clear()
        skips.clear()
        player.destroy()

        JukeBot.shardManager.getGuildById(guildId)?.audioManager?.sendingHandler = null
    }

    /*
     *  +===================+
     *  |   Player Events   |
     *  +===================+
     */
    override fun onTrackStart(player: AudioPlayer, track: AudioTrack) {
        player.isPaused = false

        if (current == null || current!!.identifier != track.identifier) {
            current = track
            val durString = if (track.info.isStream) "LIVE" else track.info.length.toTimeString()
            announce("Now Playing", "${track.info.title} - `$durString`")

            val title = track.info.title
            val nick = if (title.length > 32) "${title.substring(0, 29)}..." else title
            setNick(nick)
        }
    }

    override fun onTrackEnd(player: AudioPlayer, track: AudioTrack, endReason: AudioTrackEndReason) {
        skips.clear()
        trackPacketLost = 0
        trackPacketsSent = 0

        previous = track

        if (endReason.mayStartNext) {
            playNext()
        }
    }

    override fun onTrackException(player: AudioPlayer, track: AudioTrack, exception: FriendlyException) {
        val breadCrumb = BreadcrumbBuilder()
            .setCategory("AudioHandler")
            .setMessage("Track ID: ${track.identifier}")
            .build()

        val eventBuilder = EventBuilder().withMessage(exception.message)
            .withLevel(Event.Level.ERROR)
            .withSentryInterface(ExceptionInterface(exception))
            .withBreadcrumbs(listOf(breadCrumb))

        Sentry.capture(eventBuilder)

        if (repeat != RepeatMode.NONE)
            repeat = RepeatMode.NONE

        val problem = Helpers.rootCauseOf(exception)

//        if (banned && Database.getIsAutoPlayEnabled(guildId)) {
//            Database.setAutoPlayEnabled(guildId, false)
//        }

        announce("Playback Error", "Playback of **${track.info.title}** encountered an error!\n" +
            problem.localizedMessage)
    }

    override fun onTrackStuck(player: AudioPlayer, track: AudioTrack, thresholdMs: Long) {
        if (repeat != RepeatMode.NONE)
            repeat = RepeatMode.NONE

        announce("Track Stuck", "Playback of **${track.info.title}** has frozen and is unable to resume!")
        playNext()
    }

    /*
     *  +=======================+
     *  |   JDA Audio Sending   |
     *  +=======================+
     */
    override fun canProvide(): Boolean {
        val frameProvided = player.provide(mutableFrame)

        if (!player.isPaused) {
            if (!frameProvided) {
                trackPacketLost++
            } else {
                trackPacketsSent++
            }
        }

        return frameProvided
    }

    override fun provide20MsAudio(): ByteBuffer {
        return buffer.flip()
    }

    override fun isOpus(): Boolean {
        return true
    }

    enum class RepeatMode {
        ALL,
        SINGLE,
        NONE;

        fun humanized(): String {
            return this.toString().toLowerCase().toTitleCase()
        }
    }

    companion object {
        const val EXPECTED_PACKET_COUNT_PER_MIN = ((60 * 1000) / 20).toDouble()
    }
}
